---
title: 'LD Part 1'
subtitle: 'Access precomputed LD'
date: '2019-05-21'
slug: get-ld-remotely
code_download: true
tags: 
  - statistical genetics
  - R
  - data visualisation
draft: false
url_source: https://github.com/sinarueeger/webpage/blob/master/blogdown/content/post/2018-08-09-put-the-color-into-the-points.Rmd
url_code: https://github.com/sinarueeger/webpage/blob/master/blogdown/content/post/2018-08-09-put-the-color-into-the-points.R
header:
  caption: ''
  image: ''
output:
  blogdown::html_page:
    toc: true
---

<link href="/rmarkdown-libs/font-awesome/css/fontawesome-all.min.css" rel="stylesheet" />

<div id="TOC">
<ul>
<li><a href="#goal">Goal</a></li>
<li><a href="#two-approaches">Two approaches</a></li>
<li><a href="#our-toy-data">Our toy data</a></li>
<li><a href="#a.-a-solution-that-works-ldlink-from-nih">1A. A solution that works: ldlink from NIH</a><ul>
<li><a href="#ldproxy">LDproxy</a></li>
<li><a href="#ldmatrix">LDmatrix</a></li>
</ul></li>
<li><a href="#b.-a-solution-that-almost-works-ensembl.org">1B. A solution that almost works: ensembl.org</a><ul>
<li><a href="#what-reference-panelspopulation-can-we-choose-from">What reference panels/population can we choose from?</a></li>
<li><a href="#access-ld-between-a-snp-and-its-region">Access LD between a SNP and its region</a></li>
<li><a href="#access-ld-matrix">Access LD matrix</a></li>
<li><a href="#access-ld-between-a-snp-and-many-other-snps">Access LD between a SNP and many other SNPs</a></li>
<li><a href="#coloured-locuszoom-plot">Coloured locuszoom plot</a></li>
</ul></li>
<li><a href="#solutions-that-work-half-through">2. Solutions that work half-through</a><ul>
<li><a href="#snpsnap">SNPsnap</a></li>
<li><a href="#api-provided-by-sph.umich">API provided by sph.umich</a></li>
</ul></li>
<li><a href="#a-solution-that-does-not-work">3. A solution that does not work</a><ul>
<li><a href="#rsnpsld_search"><code>rsnps::ld_search</code></a></li>
</ul></li>
<li><a href="#conclusion">Conclusion</a></li>
<li><a href="#session-info">Session Info</a></li>
</ul>
</div>

<p>R-packages used:</p>
<pre class="r"><code>library(httr)
library(jsonlite)
library(xml2)
library(ggplot2)
theme_set(theme_bw())
library(tibble)
library(tidyr)
library(magrittr)
library(glue)
library(purrr)
library(rsnps)
library(data.table)
library(janitor)
library(stringr)
library(rsnps)</code></pre>
<p>The squared correlation between genetic markers is one way to estimate <a href="https://ldlink.nci.nih.gov/?tab=help">linkage disequilibrium</a> (LD). LD has to be computed <em>all the time</em> - either for as an input for statistical methods or to summarise results.</p>
<p>However, accessing LD estimations quickly, for a specific population and in an automated way (e.g. with R) is suprisingly difficult.</p>
<p>In this blog post I am exploring how to do this efficiently.</p>
<!---------- ------------------------------->
<div id="goal" class="section level2">
<h2>Goal</h2>
<!---------- ------------------------------->
<p>At the end of this blog post, we want to know the genetic correlation between two or more markers in a specific human population, so that we can populate the locuszoom plot from the <a href="https://sinarueeger.github.io/2018/07/30/locuszoomplot/">previous blog post</a> with coloured dots.</p>
<p>For simplicity, I will use the terms correlation, squared correlation, r, r2 and LD interchangeably.</p>
<!---------- ------------------------------->
</div>
<div id="two-approaches" class="section level2">
<h2>Two approaches</h2>
<!---------- ------------------------------->
<p>In principle, there are two ways of doing accessing LD:</p>
<ol style="list-style-type: decimal">
<li>Download (or access) the genetic data from which you want to estimate your correlations + calculate the correlations using some efficient approach.</li>
<li>Access precomputed LD estimations.</li>
</ol>
<table>
<colgroup>
<col width="9%" />
<col width="22%" />
<col width="39%" />
<col width="28%" />
</colgroup>
<thead>
<tr class="header">
<th>Approach <i class="fas  fa-wrench "></i></th>
<th>Advantages <i class="fas  fa-smile "></i></th>
<th>Downsides <i class="fas  fa-frown "></i></th>
<th>Useful when…</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>(1) <strong>Local</strong> computation of LD</td>
<td>LD matrix can be quickly updated to new reference panels</td>
<td>Requires large computation and storage space (e.g. 1000 Genomes is <a href="https://console.cloud.google.com/storage/browser/genomics-public-data/1000-genomes-phase-3/vcf/?_ga=2.227457240.-473641130.1558969048">&gt;100 few GB large</a>).</td>
<td>i) LD for a large set of SNPs is needed ii) LD from non-standard reference panel is needed.</td>
</tr>
<tr class="even">
<td>(2) <strong>Access</strong> precomputed LD</td>
<td>not need for large computation and storage space.</td>
<td>limited to certain small sets of markers, limited to possibly outdated reference panels.</td>
<td>LD for a small set of SNPs is needed</td>
</tr>
</tbody>
</table>
<p>For now, I will focus on approach (2), and then explore approach (1) in a future blog post.</p>
<p>Spoiler: Using approach (2) does not get you far. It took me quite a while to gather all the solutions that are listed below, and yet there is not one perfect/ideal solution.</p>
<!---------- ------------------------------->
</div>
<div id="our-toy-data" class="section level2">
<h2>Our toy data</h2>
<!---------- ------------------------------->
<p>We will recycle the data from the <a href="https://sinarueeger.github.io/2018/07/30/locuszoomplot/">previous blog post</a>, where the focus was on extracting annotation using the package <code>biomaRt</code>. In this blog post, we will complete that locuszoom plot by adding the LD information.</p>
<pre class="r"><code>## Data Source URL
url &lt;- &quot;https://portals.broadinstitute.org/collaboration/giant/images/2/21/BMI_All_ancestry.fmt.gzip&quot;

## Import BMI summary statistics dat.bmi &lt;- read_tsv(file = url) ##
## taking too long, let&#39;s use fread instead.

dat.bmi &lt;- data.table::fread(url, verbose = FALSE)

## Rename some columns
dat.bmi &lt;- dat.bmi %&gt;% rename(SNP = SNPNAME, P = Pvalue)

## Extract region
dat.bmi.sel &lt;- dat.bmi %&gt;% slice(which.min(P))
dat.bmi.sel

## range region
range &lt;- 5e+05
sel.chr &lt;- dat.bmi.sel$CHR
sel.pos &lt;- dat.bmi.sel$POS

data &lt;- dat.bmi %&gt;%
  filter(CHR == sel.chr, between(POS, sel.pos -
    range, sel.pos + range))

head(data)

(snp &lt;- dat.bmi.sel$SNP)</code></pre>
<p>What we are interested in is the LD between our top SNP rs1421085 and all other 82 SNPs nearby.</p>
<p>This dataset has positions on build GRCh37, while most databases are on build GRCh38 by now.</p>
<pre class="r"><code>sm &lt;- rsnps::ncbi_snp_summary(snp) %&gt;% separate(chrpos, c(&quot;chr&quot;, &quot;pos&quot;))
sel.pos == as.numeric(sm$pos)</code></pre>
<pre><code>## [1] FALSE</code></pre>
<p>Let’s quickly repeat what our primary goal is:</p>
<p>Extract the correlation between SNPs</p>
<ul>
<li>without downloading any data,</li>
<li>fairly quick and</li>
<li>in R.</li>
</ul>
<!---------- ------------------------------->
</div>
<div id="a.-a-solution-that-works-ldlink-from-nih" class="section level2">
<h2>1A. A solution that works: ldlink from NIH</h2>
<!---------- ------------------------------->
<p><a href="https://ldlink.nci.nih.gov/?tab=home">ldlink</a> is a website provided by <a href="https://dceg.cancer.gov/">NIH</a> to easily (and programmatically) request LD estimates in population groups.</p>
<p>LD is estimated from Phase 3 of the 1000 Genomes Project and super- and subpopulations can be selected.</p>
<p>There are different ways to access the LD estimations (e.g. <code>LDpair</code>, <code>LDmatrix</code>, <code>LDproxy</code>) and the same modules are also available through the <a href="https://ldlink.nci.nih.gov/?tab=help#Programatic%20Access">API</a>.</p>
<p>To access the API, you need to <a href="https://ldlink.nci.nih.gov/?tab=apiaccess">register</a> for a token (takes a few seconds).</p>
<pre class="r"><code>MYTOKEN &lt;- &quot;a_mix_of_numbers_and_characters&quot;</code></pre>
<p>Let’s look at two modules:</p>
<ul>
<li><a href="#ldproxy">LDproxy: access LD between a SNP and its region</a></li>
<li><a href="#ldmatrix">LDmatrix: access LD matrix</a></li>
</ul>
<div id="ldproxy" class="section level3">
<h3>LDproxy</h3>
<p>To get the LD between a SNP and its region.</p>
<p>First, access the API:</p>
<pre class="r"><code>LDproxy_raw &lt;- system(
  glue::glue(&quot;curl -k -X GET &#39;https://ldlink.nci.nih.gov/LDlinkRest/ldproxy?var={snp}&amp;pop=EUR&amp;r2_d=r2&amp;token={MYTOKEN}&#39;&quot;),
  intern = TRUE
)</code></pre>
<p>Then, do a bit of data wrangling to get a tidy data frame:</p>
<pre class="r"><code>LDproxy &lt;- LDproxy_raw %&gt;%
  purrr::map(., function(x) stringr::str_split(x, &quot;\t&quot;) %&gt;%
      unlist()) %&gt;% ## remove all the tabs
  do.call(rbind, .) %&gt;% ## turn into a matrix
  data.frame() %&gt;% ## turn into a data frame
  janitor::row_to_names(1) %&gt;% ## make the first row the column names
  rename(SNP = RS_Number) %&gt;% ## rename RS_Number as SNP
  mutate_at(vars(MAF:R2), function(x) as.numeric(as.character(x))) %&gt;% ## turn MAF:R2 columns numeric
  mutate(SNP = as.character(SNP)) ## turn SNP from a factor into a character
head(LDproxy)</code></pre>
<pre><code>##          SNP          Coord Alleles    MAF Distance Dprime     R2
## 1  rs1421085 chr16:53800954   (T/C) 0.4324        0 1.0000 1.0000
## 2 rs11642015 chr16:53802494   (C/T) 0.4324     1540 1.0000 1.0000
## 3 rs62048402 chr16:53803223   (G/A) 0.4324     2269 1.0000 1.0000
## 4  rs1558902 chr16:53803574   (T/A) 0.4324     2620 1.0000 1.0000
## 5 rs55872725 chr16:53809123   (C/T) 0.4324     8169 1.0000 1.0000
## 6 rs56094641 chr16:53806453   (A/G) 0.4344     5499 0.9959 0.9839
##   Correlated_Alleles RegulomeDB Function
## 1            T=T,C=C          5       NA
## 2            T=C,C=T          4       NA
## 3            T=G,C=A          5       NA
## 4            T=T,C=A          7       NA
## 5            T=C,C=T          4       NA
## 6            T=A,C=G          6       NA</code></pre>
<p>Next, join the original summary stats data with the <code>LDproxy</code> data frame.</p>
<pre class="r"><code>data_ldproxy &lt;- data %&gt;%
  right_join(LDproxy, by = c(&quot;SNP&quot; = &quot;SNP&quot;))</code></pre>
<p>Lastly, plot the summary statistics with the point colour indicating the R2.</p>
<pre class="r"><code>ggplot(data = data_ldproxy) +
  geom_point(aes(POS, -log10(P), color = R2), shape = 16) +
  labs(
    title = &quot;Locuszoom plot for BMI GWAS&quot;,
    subtitle = paste(
      &quot;Summary statistics for chromosome&quot;, sel.chr, &quot;from&quot;,
      format((sel.pos - range), big.mark = &quot;&#39;&quot;), &quot;to&quot;,
      format((sel.pos + range), big.mark = &quot;&#39;&quot;), &quot;bp&quot;
    ),
    caption = paste(&quot;Data source:&quot;, url)
  ) +
  geom_point(
    data = data_ldproxy %&gt;% filter(SNP == snp),
    aes(POS, -log10(P)), color = &quot;black&quot;, shape = 16
  ) +
  scale_color_distiller(&quot;R2 (LDproxy)&quot;,
    type = &quot;div&quot;, palette = &quot;Spectral&quot;,
    limits = c(0, 1)
  )</code></pre>
<p><img src="/post/2018-08-09-put-the-color-into-the-points_files/figure-html/plot-summarystats-ldlink-1.png" width="672" /></p>
</div>
<div id="ldmatrix" class="section level3">
<h3>LDmatrix</h3>
<p><code>LDmatrix</code> module accesses the pairwise LD between a set of SNPs.</p>
<p>Again, first access the API:</p>
<pre class="r"><code>snplist &lt;- data %&gt;%
  filter(str_detect(SNP, &quot;rs&quot;)) %&gt;%
  pull(SNP) %&gt;%
  paste(collapse = &quot;%0A&quot;)

LDmatrix_raw &lt;- system(
  glue::glue(&quot;curl -k -X GET &#39;https://ldlink.nci.nih.gov/LDlinkRest/ldmatrix?snps={snplist}&amp;pop=CEU%2BTSI%2BFIN%2BGBR%2BIBS&amp;r2_d=r2&amp;token={MYTOKEN}&#39;&quot;),
  intern = TRUE
)</code></pre>
<ul>
<li>If you want to access the dprime (d’) values, write <code>r2_d=d</code>.</li>
<li>If you want to access certain sub populations, let’s say CEU, TSI and FIN, concatenate them with <code>%B</code> in between: <code>CEU%2BTSI%2BFIN</code>.</li>
</ul>
<p>Then, do a little data tidying:</p>
<pre class="r"><code>LDmatrix &lt;- LDmatrix_raw %&gt;%
  purrr::map(., function(x) stringr::str_split(x, &quot;\t&quot;) %&gt;%
      unlist()) %&gt;%
  do.call(rbind, .) %&gt;% ## turn into a matrix
  data.frame() %&gt;% ## turn into a data.frame
  janitor::row_to_names(1) ## make the first line the column names

LDmatrix_long &lt;- LDmatrix %&gt;%
  gather(&quot;SNP2&quot;, &quot;R2&quot;, -RS_number) %&gt;% ## from wide to long
  rename(SNP = RS_number) %&gt;% ## rename RS_number
  mutate(R2 = as.numeric(R2)) %&gt;% ## make R2 numeric
  mutate_if(is.factor, as.character) ## make all factor columns characters</code></pre>
<pre><code>## Warning: attributes are not identical across measure variables;
## they will be dropped</code></pre>
<pre><code>## Warning: NAs introduced by coercion</code></pre>
<pre class="r"><code>head(LDmatrix_long)</code></pre>
<pre><code>##           SNP       SNP2    R2
## 1  rs61754093 rs61754093 1.000
## 2 rs181111349 rs61754093    NA
## 3 rs199662749 rs61754093    NA
## 4 rs189080082 rs61754093 0.000
## 5   rs6499548 rs61754093 0.023
## 6 rs139704369 rs61754093    NA</code></pre>
<p>Next, join the original summary stats data with the <code>LDmatrix_long</code> data frame.</p>
<pre class="r"><code>data_ldmatrix &lt;- data %&gt;%
  right_join(LDmatrix_long, by = c(&quot;SNP&quot; = &quot;SNP&quot;))</code></pre>
<p>Lastly, plot the summary statistics with the point colour indicating the R2.</p>
<pre class="r"><code>ggplot(data = data_ldmatrix %&gt;% filter(SNP2 == snp)) +
  geom_point(aes(POS, -log10(P), color = R2)) +
  labs(
    title = &quot;Locuszoom plot for BMI GWAS&quot;,
    subtitle = paste(
      &quot;Summary statistics for chromosome&quot;, sel.chr, &quot;from&quot;,
      format((sel.pos - range), big.mark = &quot;&#39;&quot;), &quot;to&quot;,
      format((sel.pos + range), big.mark = &quot;&#39;&quot;), &quot;bp&quot;
    ),
    caption = paste(&quot;Data source:&quot;, url)
  ) +
  geom_point(
    data = data_ldmatrix %&gt;% filter(SNP == snp &amp; SNP2 == snp),
    aes(POS, -log10(P)), color = &quot;black&quot;, shape = 16
  ) +
  scale_color_distiller(&quot;R2 (LDmatrix)&quot;,
    type = &quot;div&quot;, palette = &quot;Spectral&quot;,
    limits = c(0, 1)
  )</code></pre>
<p><img src="/post/2018-08-09-put-the-color-into-the-points_files/figure-html/plot-summarystats-ldlink-matrix-1.png" width="672" /></p>
<!---------- ------------------------------->
</div>
</div>
<div id="b.-a-solution-that-almost-works-ensembl.org" class="section level2">
<h2>1B. A solution that almost works: ensembl.org</h2>
<!---------- ------------------------------->
<p>The <a href="https://rest.ensembl.org">REST API of Ensembl</a> can do a lot (see options <a href="https://rest.ensembl.org/documentation/">here</a>). For example access precomputed LD. The webpage even provides R code to do so, which is from where I copied some snippets below.</p>
<p>To access the rest API at ensembl, we need the following three packages loaded.</p>
<pre class="r"><code>library(httr)
library(jsonlite)
library(xml2)</code></pre>
<div id="what-reference-panelspopulation-can-we-choose-from" class="section level3">
<h3>What reference panels/population can we choose from?</h3>
<!------------------------------>
<p>Currently, the largest and hence most popular public reference panel is <a href="http://www.internationalgenome.org/">1000 Genomes reference panel</a> (1KG). The 26 populations of roughly 100 individuals each can be grouped into five super populations: African (AFR), American (AMR), European (EUR), South Asian (SAS), East Asian (EAS).</p>
<p>We can ask the ENSEMBL API from what populations reference panels are available. This will return us a data frame.</p>
<pre class="r"><code>server &lt;- &quot;https://rest.ensembl.org&quot;
ext &lt;- &quot;/info/variation/populations/homo_sapiens?filter=LD&quot;

r &lt;- GET(paste(server, ext, sep = &quot;&quot;), content_type(&quot;application/json&quot;))

stop_for_status(r)

head(fromJSON(toJSON(content(r))))</code></pre>
<pre><code>##                                                  description size
## 1                              African Caribbean in Barbados   96
## 2                           African Ancestry in Southwest US   61
## 3                                      Bengali in Bangladesh   86
## 4                        Chinese Dai in Xishuangbanna, China   93
## 5 Utah residents with Northern and Western European ancestry   99
## 6                               Han Chinese in Bejing, China  103
##                      name
## 1 1000GENOMES:phase_3:ACB
## 2 1000GENOMES:phase_3:ASW
## 3 1000GENOMES:phase_3:BEB
## 4 1000GENOMES:phase_3:CDX
## 5 1000GENOMES:phase_3:CEU
## 6 1000GENOMES:phase_3:CHB</code></pre>
<p><code>name</code> stands for the population identifier. <code>size</code> refers to the number of individuals in the reference panel. Note that these are all populations with around 100 individuals (the correlation estimation will have an error that scales with the sample size). There are also the five super population available (although not listed here), simply replace the last three characters in <code>name</code> by <code>EUR</code>, <code>AFR</code>, <code>AMR</code>, <code>EAS</code>, <code>SAS</code>.</p>
<iframe seamless src="/post/2018-08-09-put-the-color-into-the-points/map-1000genomes-populations.html" width="100%" height="500">
</iframe>
<p><sup>(From <a href="../1kgmap.html">this</a> blog post.)</sup></p>
<p>We want the LD information, so that we can add this info to the locuszoom plot. But how do we know which population to pick? One way is to read up what kind of individuals were present. In our case - mostly Europeans (<code>EUR</code>). But we could also build some pooled LD matrix of different populations.</p>
<p>Now that we know which reference panel we want to use, we can use the different rest APIs.</p>
<ul>
<li><a href="#access-ld-between-a-snp-and-its-region">Access LD between a SNP and its region</a></li>
<li><a href="#access-ld-matrix">Access LD matrix</a></li>
<li><a href="#access-ld-between-a-snp-and-many-other-snps">Access LD between a SNP and many other SNPs</a></li>
</ul>
</div>
<div id="access-ld-between-a-snp-and-its-region" class="section level3">
<h3>Access LD between a SNP and its region</h3>
<!------------------------------>
<p>This API is described <a href="https://rest.ensembl.org/documentation/info/ld_id_get">here</a>.</p>
<p>The default window size is 500 kb. There are also thresholds for <code>r2</code> (e.g. if you want to filter all SNPs with an <code>r2 &gt; 0.8</code>).</p>
<p>The only input required is the SNP rsid, marked with <code>{snp}</code>.</p>
<pre class="r"><code>snp</code></pre>
<pre><code>## [1] &quot;rs1421085&quot;</code></pre>
<pre class="r"><code>server &lt;- &quot;https://rest.ensembl.org&quot;
ext &lt;- glue::glue(&quot;/ld/human/{snp}/1000GENOMES:phase_3:EUR?&quot;)
## Window size in kb. The maximum allowed value for the window size is 500 kb. LD is computed for the given variant and all variants that are located within the specified window.

r &lt;- GET(paste(server, ext, sep = &quot;&quot;), content_type(&quot;application/json&quot;))
stop_for_status(r)
LD.SNP.region &lt;- as_tibble(fromJSON(toJSON(content(r)))) %&gt;%
  unnest() %&gt;%
  mutate(r2 = as.numeric(r2))</code></pre>
<pre><code>## Warning: `cols` is now required.
## Please use `cols = c(variation1, d_prime, r2, population_name, variation2)`</code></pre>
<pre class="r"><code>head(LD.SNP.region)</code></pre>
<pre><code>## # A tibble: 6 x 5
##   variation1 d_prime      r2 population_name         variation2
##   &lt;chr&gt;      &lt;chr&gt;     &lt;dbl&gt; &lt;chr&gt;                   &lt;chr&gt;     
## 1 rs1421085  0.327013 0.0834 1000GENOMES:phase_3:EUR rs8043738 
## 2 rs1421085  0.992084 0.409  1000GENOMES:phase_3:EUR rs2042031 
## 3 rs1421085  0.846280 0.647  1000GENOMES:phase_3:EUR rs11642841
## 4 rs1421085  1.000000 0.957  1000GENOMES:phase_3:EUR rs9940128 
## 5 rs1421085  0.999948 0.0552 1000GENOMES:phase_3:EUR rs73612011
## 6 rs1421085  0.907868 0.648  1000GENOMES:phase_3:EUR rs8057044</code></pre>
<p>As a result, <code>LD.snp.region</code> contains the <code>r2</code> of our top SNP with all SNPs that were +/- 500 kb away.</p>
<p>What if we want the correlation between all SNPs?</p>
</div>
<div id="access-ld-matrix" class="section level3">
<h3>Access LD matrix</h3>
<!------------------------------>
<p>For this, we need the rest API <a href="https://rest.ensembl.org/documentation/info/ld_region_get">here</a>.</p>
<p>We can calculate the LD matrix of a full region, max 1 Mb wide. For fast computation, we limit it to +/- 50 kb.</p>
<pre class="r"><code>## Query region. A maximum of 1Mb is allowed.

ext &lt;- glue::glue(&quot;/ld/human/region/{sel.chr}:{sel.pos - range/20}..{sel.pos + range/20}/1000GENOMES:phase_3:EUR?&quot;)

r &lt;- GET(paste(server, ext, sep = &quot;&quot;), content_type(&quot;application/json&quot;))
stop_for_status(r)
LD.matrix.region &lt;- as_tibble(fromJSON(toJSON(content(r)))) %&gt;%
  unnest() %&gt;%
  mutate(r2 = as.numeric(r2))</code></pre>
<pre><code>## Warning: `cols` is now required.
## Please use `cols = c(population_name, r2, variation2, variation1, d_prime)`</code></pre>
<pre class="r"><code>head(LD.matrix.region)</code></pre>
<pre><code>## # A tibble: 6 x 5
##   population_name            r2 variation2  variation1 d_prime 
##   &lt;chr&gt;                   &lt;dbl&gt; &lt;chr&gt;       &lt;chr&gt;      &lt;chr&gt;   
## 1 1000GENOMES:phase_3:EUR 0.761 rs9933509   rs8057044  0.999998
## 2 1000GENOMES:phase_3:EUR 1     rs150763868 rs72803680 1.000000
## 3 1000GENOMES:phase_3:EUR 0.106 rs7206122   rs62033401 0.999940
## 4 1000GENOMES:phase_3:EUR 0.731 rs9935401   rs8057044  0.999997
## 5 1000GENOMES:phase_3:EUR 0.106 rs141816793 rs62033399 0.999943
## 6 1000GENOMES:phase_3:EUR 0.996 rs9941349   rs9931900  1.000000</code></pre>
</div>
<div id="access-ld-between-a-snp-and-many-other-snps" class="section level3">
<h3>Access LD between a SNP and many other SNPs</h3>
<!------------------------------>
<p>The <a href="https://rest.ensembl.org/documentation/info/ld_pairwise_get">third and last option</a> is to pass on a set of SNP rs ids, and access the LD among these. Implemented in the ENSEMBL API is only the LD between two SNPs, so we will have to extend this to many SNPs.</p>
<pre class="r"><code>extract_ld &lt;- function(SNP.id2 = NULL, SNP.id1 = NULL, POP = NULL) {
  ext &lt;- glue::glue(&quot;/ld/human/pairwise/{SNP.id1}/{SNP.id2}/&quot;) ## filter POP further down

  server &lt;- &quot;https://rest.ensembl.org&quot;

  r &lt;- GET(paste(server, ext, sep = &quot;&quot;), content_type(&quot;application/json&quot;))
  stop_for_status(r)

  out &lt;- as_tibble(fromJSON(toJSON(content(r)))) %&gt;%
    unnest() %&gt;%
    filter(stringr::str_detect(population_name, POP))

  return(out)
}


## see futher down why intersect here
other.snps &lt;- intersect(LD.SNP.region$variation2, data$SNP)

## cacluate LD for all other.snps SNPs
LD.matrix.snps &lt;- purrr::map_df(other.snps, extract_ld, snp, &quot;EUR&quot;) %&gt;%
  mutate(r2 = as.numeric(r2)) %&gt;%
  bind_rows() %&gt;%
  unnest()</code></pre>
<pre><code>## Warning: `cols` is now required.
## Please use `cols = c(variation2, d_prime, variation1, r2, population_name)`

## Warning: `cols` is now required.
## Please use `cols = c(variation2, d_prime, variation1, r2, population_name)`

## Warning: `cols` is now required.
## Please use `cols = c(variation2, d_prime, variation1, r2, population_name)`</code></pre>
<pre><code>## Warning: `cols` is now required.
## Please use `cols = c(d_prime, variation1, variation2, r2, population_name)`</code></pre>
<pre><code>## Warning: `cols` is now required.
## Please use `cols = c(variation2, variation1, d_prime, population_name, r2)`</code></pre>
<pre><code>## Warning: `cols` is now required.
## Please use `cols = c(variation1, d_prime, variation2, population_name, r2)`</code></pre>
<pre><code>## Warning: `cols` is now required.
## Please use `cols = c(r2, population_name, d_prime, variation1, variation2)`</code></pre>
<pre><code>## Warning: `cols` is now required.
## Please use `cols = c(population_name, r2, variation1, d_prime, variation2)`</code></pre>
<pre><code>## Warning: `cols` is now required.
## Please use `cols = c(variation2, d_prime, variation1, r2, population_name)`</code></pre>
<pre><code>## Warning: `cols` is now required.
## Please use `cols = c(d_prime, variation1, variation2, r2, population_name)`</code></pre>
<pre><code>## Warning: `cols` is now required.
## Please use `cols = c()`</code></pre>
<pre class="r"><code>LD.matrix.snps</code></pre>
<pre><code>## # A tibble: 0 x 5
## # … with 5 variables: variation2 &lt;chr&gt;, d_prime &lt;chr&gt;, variation1 &lt;chr&gt;,
## #   r2 &lt;dbl&gt;, population_name &lt;chr&gt;</code></pre>
<p>Calculate the LD matrix (<code>LD.matrix.region</code>) or the LD between SNP pairs (<code>LD.matrix.snps</code>) takes <em>a lot</em> of time!</p>
</div>
<div id="coloured-locuszoom-plot" class="section level3">
<h3>Coloured locuszoom plot</h3>
<!---------- ------------------------------->
<p>For the locuszoom plot we need only the correlation between the top SNP and all other SNPs. So we join the object <code>LD.SNP.region</code> to <code>data</code>.</p>
<pre class="r"><code>data_ensembl &lt;- data %&gt;%
  full_join(LD.SNP.region, by = c(&quot;SNP&quot; = &quot;variation2&quot;))</code></pre>
<pre class="r"><code>ggplot(data = data_ensembl) +
  geom_point(aes(POS, -log10(P), color = r2), shape = 16) +
  labs(
    title = &quot;Locuszoom plot for BMI GWAS&quot;,
    subtitle = paste(&quot;Summary statistics for chromosome&quot;, sel.chr, &quot;from&quot;, format((sel.pos - range), big.mark = &quot;&#39;&quot;), &quot;to&quot;, format((sel.pos + range), big.mark = &quot;&#39;&quot;), &quot;bp&quot;),
    caption = paste(&quot;Data source:&quot;, url)
  ) +
  geom_point(
    data = data_ensembl %&gt;% filter(SNP == &quot;rs1421085&quot;),
    aes(POS, -log10(P)), color = &quot;black&quot;, shape = 16
  ) +
  scale_color_distiller(&quot;R2 (ensembl)&quot;, type = &quot;div&quot;, palette = &quot;Spectral&quot;, limits = c(0, 1))</code></pre>
<pre><code>## Warning: Removed 205 rows containing missing values (geom_point).</code></pre>
<p><img src="/post/2018-08-09-put-the-color-into-the-points_files/figure-html/plot-summarystats-ld-1.png" width="672" /></p>
<!-- REST API: https://rviews.rstudio.com/2018/07/23/rest-apis-and-plumber/ ----->
<!---------- ------------------------------->
</div>
</div>
<div id="solutions-that-work-half-through" class="section level2">
<h2>2. Solutions that work half-through</h2>
<!---------- ------------------------------->
<div id="snpsnap" class="section level3">
<h3>SNPsnap</h3>
<!------------------------------>
<ul>
<li>SNPsnap: <a href="https://data.broadinstitute.org/mpg/snpsnap/database_download.html" class="uri">https://data.broadinstitute.org/mpg/snpsnap/database_download.html</a></li>
<li>uses a limited set of 1KG populations (EUR, EAS, WAFR).</li>
</ul>
</div>
<div id="api-provided-by-sph.umich" class="section level3">
<h3>API provided by sph.umich</h3>
<!------------------------------>
<ul>
<li><a href="http://portaldev.sph.umich.edu/docs/api/v1/?shell#single-variant-statistics">API</a></li>
<li>uses limited set of 1KG populations (ALL, EUR)</li>
<li>see <a href="https://github.com/statgen/locuszoom-api/issues/21#issuecomment-414434704">github issue</a></li>
</ul>
<!---- ### Using locuszoom software ----->
<!------------------------------>
<!---- One option (or rather a hack) would be to use [locuszoom software](https://genome.sph.umich.edu/wiki/LocusZoom_Standalone) to extract their LD information. However, locuszoom comes with database and LD files, [yielding up to 23 G](https://github.com/statgen/locuszoom-standalone). Besides, its not maintained anymore, since a [new version](https://github.com/statgen/locuszoom) for [interactive usage](http://locuszoom.org/locuszoomjs.php) was developed. 

------->
<!---------- ------------------------------->
</div>
</div>
<div id="a-solution-that-does-not-work" class="section level2">
<h2>3. A solution that does not work</h2>
<!---------- ------------------------------->
<div id="rsnpsld_search" class="section level3">
<h3><code>rsnps::ld_search</code></h3>
<!------------------------------>
<p>A perfect solution would have been the function <code>ld_search</code> from R package <a href="ropensci%20webpage"><code>rsnps</code></a>. It has arguments to choose the reference panel, the population, the distance from the focal SNP.</p>
<p>The problem is that it only uses old reference panels (HapMap and 1KG-phase1). Meaning, many newer reference panel populations are left out.</p>
<p>But the main problem is, that the broad institute has taken down the <a href="https://www.broadinstitute.org/snap/snap">snap server</a> that <code>ld_search</code> used to access (see <a href="https://github.com/ropensci/rsnps/issues/60">github issue</a>); hence <code>ld_search</code> <a href="https://github.com/ropensci/rsnps/blob/master/R/LDSearch.R">is defunct</a>.</p>
<!---------- ------------------------------->
</div>
</div>
<div id="conclusion" class="section level2">
<h2>Conclusion</h2>
<!---------- ------------------------------->
<p>The
<a href="https://ldlink.nci.nih.gov/?tab=home">ldlink</a> API with the <code>LDproxy</code> module seems the most perfect solution for now.</p>
<p>This will probably change with changing technology and larger reference panels.</p>
</div>
<div id="session-info" class="section level2">
<h2>Session Info</h2>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>## R version 3.5.1 (2018-07-02)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: macOS  10.14.5
## 
## Matrix products: default
## BLAS: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] janitor_1.2.0      data.table_1.12.2  rsnps_0.3.2.9121  
##  [4] glue_1.3.1.9000    magrittr_1.5       xml2_1.2.0        
##  [7] jsonlite_1.6       httr_1.4.0         forcats_0.4.0.9000
## [10] stringr_1.4.0      dplyr_0.8.3.9000   purrr_0.3.2       
## [13] readr_1.3.1        tidyr_0.8.3.9000   tibble_2.1.3      
## [16] ggplot2_3.2.0.9000 tidyverse_1.2.1   
## 
## loaded via a namespace (and not attached):
##  [1] tidyselect_0.2.5 xfun_0.8         haven_2.1.1      lattice_0.20-38 
##  [5] colorspace_1.4-1 vctrs_0.2.0.9000 generics_0.0.2   htmltools_0.3.6 
##  [9] yaml_2.2.0       utf8_1.1.4       XML_3.98-1.20    rlang_0.4.0     
## [13] pillar_1.4.2     httpcode_0.2.0   withr_2.1.2      modelr_0.1.4    
## [17] readxl_1.3.1     plyr_1.8.4       munsell_0.5.0    blogdown_0.13   
## [21] gtable_0.3.0     cellranger_1.1.0 rvest_0.3.4      evaluate_0.14   
## [25] knitr_1.23       curl_3.3         fansi_0.4.0      triebeard_0.3.0 
## [29] urltools_1.7.3   broom_0.5.2      Rcpp_1.0.1       scales_1.0.0    
## [33] backports_1.1.4  hms_0.4.2        digest_0.6.20    stringi_1.4.3   
## [37] bookdown_0.11    grid_3.5.1       cli_1.1.0        tools_3.5.1     
## [41] lazyeval_0.2.2   crul_0.8.0       crayon_1.3.4     pkgconfig_2.0.2 
## [45] zeallot_0.1.0    ellipsis_0.2.0.1 lubridate_1.7.4  assertthat_0.2.1
## [49] rmarkdown_1.13   rstudioapi_0.10  R6_2.4.0         icon_0.1.0      
## [53] nlme_3.1-140     compiler_3.5.1</code></pre>
</div>
